### ISSUE TITLE
babel-preset-react-native@5.0.2 and HOC

### ISSUE THREAD

**Issue by morten-olsen** [facebook/react-native#20520]
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
  React Native Environment Info:
    System:
      OS: macOS High Sierra 10.13.6
      CPU: x64 Intel(R) Core(TM) m3-6Y30 CPU @ 0.90GHz
      Memory: 33.01 MB / 8.00 GB
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 9.4.0 - ~/.build-tools/nvm/versions/node/v9.4.0/bin/node
      Yarn: 1.7.0 - /usr/local/bin/yarn
      npm: 5.6.0 - ~/.build-tools/nvm/versions/node/v9.4.0/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 11.4, macOS 10.13, tvOS 11.4, watchOS 4.3
    IDEs:
      Android Studio: 3.1 AI-173.4819257
      Xcode: 9.4/9F1027a - /usr/bin/xcodebuild
    npmPackages:
      react: 16.4.1 => 16.4.1
      react-native: 0.56.0 => 0.56.0

## Description

0
down vote
favorite
I have just updated a project from RN 53 to 56, which means that babel also has been updated from version 6 to 7, which is causing me an issue which I can't really make sense of.

It appears that classes extended inside a function retains only the class they extends, but not the extentions. The new constructor appears to be run though.

The issue seams to be caused by `babel-preset-react-native@5.0.2`

Below is a small demo for reproducing the issue

If i run this as `yarn babel-node --presets=react-native index.js` `new CB().b` never gets defined

```
a:  function
b:  undefined
c:  string
```

If i run it without presets or `@babel/preset-env` it behaves as expected

```
a:  function
b:  function
c:  string
```

## Reproducible Demo

**index.js**
```javascript
class A {
  a() {
    return 'a';
  }
}

const createB = () => {
  class B extends A {
    constructor() {
      super();
      this.c = 'c';
    }

    b() {
      return 'b';
    }
  }

  return B;
}

const CB = createB();

console.log('a: ', typeof new CB().a);
console.log('b: ', typeof new CB().b);
console.log('c: ', typeof new CB().c);
```

**package.json**
```
{
  "name": "err",
  "version": "1.0.0",
  "devDependencies": {
    "@babel/cli": "^7.0.0-beta.55",
    "@babel/core": "^7.0.0-beta.55",
    "@babel/node": "^7.0.0-beta.55",
    "@babel/preset-env": "^7.0.0-beta.55",
    "babel-preset-react-native": "^5.0.2"
  }
}
```

**Comment by kelset** [comment]
Yes, we upgraded to babel7, you can find the details in the changelog and in particular it looks like you have not fixed the deps to 47 as we wrote [here](https://github.com/react-native-community/react-native-releases/blob/master/CHANGELOG.md#react-native-now-uses-babel-7).

Can you check again after fixing your dependencies?
